### PR TITLE
feat: `Index` and `IndexMut` traits

### PR DESCRIPTION
**Pull request by JonathanStarup**
Part of #7983

Manual Test
```scala
def main(): Unit \ IO = region rc {
    use Index.get;
    use IndexMut.put;
    let l = "a" :: "b" :: "c" :: Nil;
    let v = l |> List.toVector;
    let a = l |> List.toArray(rc);
    let m = l |> List.zipWithIndex |> List.toMap |> Map.toMutMap(rc);
    testIt(() -> get(1, v));
    testIt(() -> get(3, v));

    println("");
    testIt(() -> get(1, a));
    testIt(() -> put(1, "z", a));
    testIt(() -> get(1, a));
    testIt(() -> get(4, a));
    testIt(() -> OutOfBounds.asRes(get(4))(a));

    println("");
    testIt(() -> get(0, m));
    testIt(() -> put(0, "q", m));
    testIt(() -> get(0, m));
    testIt(() -> get(4, m));
    ()
}

def testIt(f: Unit -> a \ ef): Unit \ IO + ((ef - KeyNotFound) - OutOfBounds) with ToString[a] = {
    match KeyNotFound.asRes(OutOfBounds.asRes(f))() |> Result.flatten {
        case Ok(v) => printUnlessUnit(v)
        case Err(msg) => printUnlessUnit(msg)
    }
}
```
Output
```
b
index 3 is out of bounds for Vector of length 3

b
z
index 4 is out of bounds for Array of length 3
Err(index 4 is out of bounds for Array of length 3)

a
q
key not found
```

### CODE REVIEW

**Comment by JonathanStarup** [comment]
@mlutze can you guess why this occurs?
```
>> Mismatched signature 'put' required by 'IndexMut'.

40 |     pub def put(i: k, v: v, t: MutMap[k, v, r]): Unit \ r = MutMap.put!(i, v, t)
                 ^^^
                 mismatched signature.

Expected scheme: ∀(k, v, r). Idx[MutMap[k, v, r]] -> (Elm[MutMap[k, v, r]] -> (MutMap[k, v, r] -> Unit \ Aef[MutMap[k, v, r]])) with IndexMut[MutMap[k, v, r]]
Actual scheme:   k -> (v -> (MutMap[k, v, r] -> Unit \ r)) with IndexMut[MutMap[k, v, r]]
```

**Comment by mlutze** [comment]
> @mlutze can you guess why this occurs? It looks like the same types to me based on the associated types
> 
> ```
> >> Mismatched signature 'put' required by 'IndexMut'.
> 
> 40 |     pub def put(i: k, v: v, t: MutMap[k, v, r]): Unit \ r = MutMap.put!(i, v, t)
>                  ^^^
>                  mismatched signature.
> 
> Expected scheme: ∀(k, v, r). Idx[MutMap[k, v, r]] -> (Elm[MutMap[k, v, r]] -> (MutMap[k, v, r] -> Unit \ Aef[MutMap[k, v, r]])) with IndexMut[MutMap[k, v, r]]
> Actual scheme:   k -> (v -> (MutMap[k, v, r] -> Unit \ r)) with IndexMut[MutMap[k, v, r]]
> ```

I'd guess the solver is broken...

**Comment by magnus-madsen** [comment]
frowning-sad-face-large-size-of-yellow-emoji-smile-free-vector

**Comment by magnus-madsen** [comment]
Is this a rigidity issue?

**Comment by JonathanStarup** [comment]
maybe? since the second the type is not quantified and the first is

File: main/src/library/IndexMut.flix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    type Elm: Type
+
+    ///
+    /// The associated effect of the data structure which represents the effect
+    /// of inserting elements.
+    ///
+    type Aef: Eff
+
+    /// Insert the element `v` at index `i` in `t`.
+    pub def put(i: Index.Idx[t], v: Index.Elm[t], t: t): Unit \ Index.Aef[t]
```
Comment (mlutze) [inline main/src/library/IndexMut.flix:34]: Should this be `IndexMut.Aef`?

Comment (mlutze) [reply main/src/library/IndexMut.flix:34]: Same for the others

File: main/src/library/InvalidKey.flix [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// Indicates that a key was invalid.
+/// Useful for instances of `Index` or `IndexMut`.
+///
+eff InvalidKey {
```
Comment (magnus-madsen) [inline main/src/library/InvalidKey.flix:21]: The key is not invalid. Rather it does not exist: I would propose we call it either "NoSuchElement" or "InvalidIndex" or "IndexOutOfBounds". I think I prefer the last as the most precise.

File: main/src/library/InvalidKey.flix [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    /// Stops execution with an error message.
+    pub def error(msg: String): Void
```
Comment (magnus-madsen) [inline main/src/library/InvalidKey.flix:24]: I think, when there is one operation, the effect and operations should have the same name. Then, later, we can add sugar.

File: main/src/library/InvalidKey.flix [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+}
+
+mod InvalidKey {
+
+    /// Removes the `InvalidKey` effect from a function `f`, if `f` `do` `InvalidKey`.
+    pub def asRes(f: a -> b \ ef): (a -> Result[String, b] \ ef - InvalidKey) = {
```
Comment (magnus-madsen) [inline main/src/library/InvalidKey.flix:31]: 😮

File: main/src/library/MutMap.flix [hunk 5]
```diff
@@ -21,6 +21,23 @@ pub enum MutMap[k: Type, v: Type, r: Region] {
     case MutMap(Region[r], Ref[Map[k, v], r])
 }
 
+instance Index[MutMap[k, v, r]] with Order[k] {
+    type Idx = k
+    type Elm = v
+    type Aef = r + InvalidKey
+    pub def get(i: k, t: MutMap[k, v, r]): v \ r + InvalidKey = {
```
Comment (magnus-madsen) [inline main/src/library/MutMap.flix:28]: I think a linebreak after associated types/effects is reasonable?

File: main/src/library/InvalidKey.flix (hunk 2 above)
Comment (JonathanStarup) [reply main/src/library/InvalidKey.flix:21]: You're reading it as "key is not valid (forever)" but I think of it as "key is not valid (right now)". I think youre right that `InvalidKey` would be a better fit, compared to `InvalidIndex` if it was also called `Mapping` and `MappingMut`.

I don't think out of bounds is good, it sounds nonsense when you're working on `Map` fx

Another option is to have both `OutOfBounds` and `InvalidIndex`.

Comment (magnus-madsen) [reply main/src/library/InvalidKey.flix:21]: For Java familiarity, I think OutOfBounds would be most understandable?

**Comment by magnus-madsen** [comment]
@JonathanStarup If you fix this first thing, I can have @sockmaster27 add instances.

Do we have a list of instances?

File: main/src/library/InvalidKey.flix (hunk 2 above)
Comment (mlutze) [reply main/src/library/InvalidKey.flix:21]: `OutOfBounds` feels very weird to me for maps. What are the bounds?

Comment (magnus-madsen) [reply main/src/library/InvalidKey.flix:21]: The set of keys. Its just a collection of discrete bounds.

File: main/src/library/MutMap.flix (hunk 5 above)
Comment (JonathanStarup) [reply main/src/library/MutMap.flix:32]: I did it now but I'm not a fan

**Comment by JonathanStarup** [comment]
@magnus-madsen im happy now, ill make a list of instances to do

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**